metricbeat/module/prometheus/remote_write/data.go

```diff
@@ -27,18 +27,24 @@ import (
 )
 
 // DefaultRemoteWriteEventsGeneratorFactory returns the default prometheus events generator
-func DefaultRemoteWriteEventsGeneratorFactory(ms mb.BaseMetricSet) (RemoteWriteEventsGenerator, error) {
-	return &remoteWriteEventGenerator{}, nil
+func DefaultRemoteWriteEventsGeneratorFactory(ms mb.BaseMetricSet, countMetrics bool) (RemoteWriteEventsGenerator, error) {
```

Comment:
Or you could use the Functional Option pattern like as follows:

``` go
type RemoteWriteEventsGeneratorOption func(*remoteWriteEventGenerator)

func WithCountMetrics(countMetrics bool) RemoteWriteEventsGeneratorOption {
    return func(r *remoteWriteEventGenerator) {
        r.countMetrics = countMetrics
    }
}

func DefaultRemoteWriteEventsGeneratorFactory(ms mb.BaseMetricSet, opts ...RemoteWriteEventsGeneratorOption) (RemoteWriteEventsGenerator, error) {
    generator := &remoteWriteEventGenerator{}
    for _, opt := range opts {
        opt(generator)
    }
    return generator, nil
}

```
And then use it like:

`generator, err := DefaultRemoteWriteEventsGeneratorFactory(ms, WithCountMetrics(true))`

Reply:
This is a very good point. I'll do this.

Reply:
Okay, Aman I tried it. Although adding in metricbeat module is fine, here the complicated part is the x-pack metricbeat. With the FO pattern, not able to get the flow correct in x-pack part of remote write.

While working on this, came across more ways to make it simpler and better (even without FO pattern) but that too did not work. So, MetricSetBuilder is the place where we can do it pretty easily if we'd have defined both the struct types that implement `PromEventsGenerator` in metricbeat module and not separately i.e., different in metricbeat and x-pack metricbeat. Cannot even import the type as it would create cyclic dependency A calls B, B calls A.

If the remote_write code was written similarly to that of collector, then also it could've been possible. I hope I didn't miss anything, let me know if you still think there is a way to add FO pattern in a simple, clean way.

Reply:
@shmsr I tried the changes I suggested. I was able to get it to work using the FO pattern. Here's the diff patch of the changes for you to refer to.  See if these changes look good and can be incorporated. 

```diff
diff --git a/metricbeat/module/prometheus/remote_write/data.go b/metricbeat/module/prometheus/remote_write/data.go
index 6fd57358bb..fa2b4796b2 100644
--- a/metricbeat/module/prometheus/remote_write/data.go
+++ b/metricbeat/module/prometheus/remote_write/data.go
@@ -27,18 +27,30 @@ import (
 )
 
 // DefaultRemoteWriteEventsGeneratorFactory returns the default prometheus events generator
-func DefaultRemoteWriteEventsGeneratorFactory(ms mb.BaseMetricSet, metricsCount bool) (RemoteWriteEventsGenerator, error) {
-	return &remoteWriteEventGenerator{metricsCount: metricsCount}, nil
+func DefaultRemoteWriteEventsGeneratorFactory(ms mb.BaseMetricSet, opts ...RemoteWriteEventsGeneratorOption) (RemoteWriteEventsGenerator, error) {
+	generator := &RemoteWriteEventGenerator{}
+	for _, opt := range opts {
+		opt(generator)
+	}
+
+	return generator, nil
 }
 
-type remoteWriteEventGenerator struct {
+type RemoteWriteEventGenerator struct {
 	metricsCount bool
 }
 
-func (p *remoteWriteEventGenerator) Start() {}
-func (p *remoteWriteEventGenerator) Stop()  {}
+type RemoteWriteEventsGeneratorOption func(r *RemoteWriteEventGenerator)
+
+func WithCountMetrics(countMetrics bool) RemoteWriteEventsGeneratorOption {
+	return func(r *RemoteWriteEventGenerator) {
+		r.metricsCount = countMetrics
+	}
+}
+func (p *RemoteWriteEventGenerator) Start() {}
+func (p *RemoteWriteEventGenerator) Stop()  {}
 
-func (p *remoteWriteEventGenerator) GenerateEvents(metrics model.Samples) map[string]mb.Event {
+func (p *RemoteWriteEventGenerator) GenerateEvents(metrics model.Samples) map[string]mb.Event {
 	eventList := map[string]mb.Event{}
 
 	for _, metric := range metrics {
diff --git a/metricbeat/module/prometheus/remote_write/remote_write.go b/metricbeat/module/prometheus/remote_write/remote_write.go
index 31d119e9d0..95048f0526 100644
--- a/metricbeat/module/prometheus/remote_write/remote_write.go
+++ b/metricbeat/module/prometheus/remote_write/remote_write.go
@@ -52,7 +52,7 @@ type RemoteWriteEventsGenerator interface {
 }
 
 // RemoteWriteEventsGeneratorFactory creates a RemoteWriteEventsGenerator when instanciating a metricset
-type RemoteWriteEventsGeneratorFactory func(ms mb.BaseMetricSet, metricsCount bool) (RemoteWriteEventsGenerator, error)
+type RemoteWriteEventsGeneratorFactory func(ms mb.BaseMetricSet, opts ...RemoteWriteEventsGeneratorOption) (RemoteWriteEventsGenerator, error)
 
 type MetricSet struct {
 	mb.BaseMetricSet
@@ -69,7 +69,7 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
 		return nil, err
 	}
 
-	promEventsGen, err := DefaultRemoteWriteEventsGeneratorFactory(base, config.MetricsCount)
+	promEventsGen, err := DefaultRemoteWriteEventsGeneratorFactory(base, WithCountMetrics(config.MetricsCount))
 	if err != nil {
 		return nil, err
 	}
@@ -99,7 +99,7 @@ func MetricSetBuilder(genFactory RemoteWriteEventsGeneratorFactory) func(base mb
 			return nil, err
 		}
 
-		promEventsGen, err := genFactory(base, config.MetricsCount)
+		promEventsGen, err := genFactory(base, WithCountMetrics(config.MetricsCount))
 		if err != nil {
 			return nil, err
 		}
diff --git a/metricbeat/module/prometheus/remote_write/remote_write_test.go b/metricbeat/module/prometheus/remote_write/remote_write_test.go
index f6f7b03eda..fc00754997 100644
--- a/metricbeat/module/prometheus/remote_write/remote_write_test.go
+++ b/metricbeat/module/prometheus/remote_write/remote_write_test.go
@@ -28,7 +28,7 @@ import (
 
 // TestGenerateEventsCounter tests counter simple cases
 func TestGenerateEventsCounter(t *testing.T) {
-	g := remoteWriteEventGenerator{}
+	g := RemoteWriteEventGenerator{}
 
 	timestamp := model.Time(424242)
 	timestamp1 := model.Time(424243)
@@ -237,7 +237,7 @@ func TestMetricsCount(t *testing.T) {
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			generator := remoteWriteEventGenerator{
+			generator := RemoteWriteEventGenerator{
 				metricsCount: true,
 			}
 
diff --git a/x-pack/metricbeat/module/prometheus/remote_write/config.go b/x-pack/metricbeat/module/prometheus/remote_write/config.go
index 9e86facb1d..0f0d91ead7 100644
--- a/x-pack/metricbeat/module/prometheus/remote_write/config.go
+++ b/x-pack/metricbeat/module/prometheus/remote_write/config.go
@@ -10,6 +10,7 @@ import (
 )
 
 type config struct {
+	MetricsCount  bool          `config:"metrics_count"`
 	UseTypes      bool          `config:"use_types"`
 	RateCounters  bool          `config:"rate_counters"`
 	TypesPatterns TypesPatterns `config:"types_patterns" yaml:"types_patterns,omitempty"`
diff --git a/x-pack/metricbeat/module/prometheus/remote_write/data.go b/x-pack/metricbeat/module/prometheus/remote_write/data.go
index bf91a10a70..4c946da5e7 100644
--- a/x-pack/metricbeat/module/prometheus/remote_write/data.go
+++ b/x-pack/metricbeat/module/prometheus/remote_write/data.go
@@ -36,7 +36,7 @@ type histogram struct {
 	metricName string
 }
 
-func remoteWriteEventsGeneratorFactory(base mb.BaseMetricSet, metricsCount bool) (rw.RemoteWriteEventsGenerator, error) {
+func remoteWriteEventsGeneratorFactory(base mb.BaseMetricSet, opts ...rw.RemoteWriteEventsGeneratorOption) (rw.RemoteWriteEventsGenerator, error) {
 	config := defaultConfig
 	if err := base.Module().UnpackConfig(&config); err != nil {
 		return nil, err
@@ -51,7 +51,7 @@ func remoteWriteEventsGeneratorFactory(base mb.BaseMetricSet, metricsCount bool)
 		g := remoteWriteTypedGenerator{
 			counterCache: counters,
 			rateCounters: config.RateCounters,
-			metricsCount: metricsCount,
+			metricsCount: config.MetricsCount,
 		}
 
 		var err error
@@ -67,7 +67,7 @@ func remoteWriteEventsGeneratorFactory(base mb.BaseMetricSet, metricsCount bool)
 		return &g, nil
 	}
 
-	return rw.DefaultRemoteWriteEventsGeneratorFactory(base, metricsCount)
+	return rw.DefaultRemoteWriteEventsGeneratorFactory(base, opts...)
 }
```
**_Note: These changes are just to demonstrate what I suggested in the comment up above. If you think these changes break the functionality, feel free to ignore them in favour of correctness._**

Reply:
Yes, thank you!! This helps! I'll apply this patch!

Reply:
Done. Please check!